### PR TITLE
fix: Unbreak FreeBSD builds by implementing find_debug_servers()

### PR DESCRIPTION
**Pull request by herrhotzenplotz**
I tried building quick-lint for testing purposes. For some reason it went belly-up because `find_debug_servers()` was undefined. Since I didn't want to just ifdef it out I figured I could just implement it.

Tests seem to pass locally on FreeBSD 13.1-RELEASE-p7 amd64. I will test on arm64 as well (and report back if it fails).

Lemme know if you find any issues with my implementation.

### Same host

``` console
$ build-native/quick-lint-js --lsp
note: quick-lint-js debug server started at http://127.0.0.1:37653/
```

And then in a different terminal:

``` console
$ build-native/quick-lint-js --debug-apps
PROCESS ID  SERVER URL
11346       http://localhost:37653/
$
```

### Server running in a Jail (see comment in code)

``` console
# jail -c name=blah ip4.addr="re0|192.168.2.96/24" path=/ exec.start='su nico -c "/usr/home/nico/src/quick-lint-js/build-native/quick-lint-js --lsp"'
note: quick-lint-js debug server started at http://192.168.2.96:52036/
```

In this case we get an empty response:

``` console
$ build-native/quick-lint-js --debug-apps
PROCESS ID  SERVER URL
$
```

However, the quick-lint-js debug server is still reachable from the printed IP address.

### CODE REVIEW

**Comment by strager** [comment]
> For some reason it went belly-up because `find_debug_servers()` was undefined.

This doesn't make sense for the default configuration. We have a default implementation at the bottom of the file:
https://github.com/quick-lint/quick-lint-js/blob/5af79aab68148a84ae9f5452b051a7e78357e348/src/quick-lint-js/debug/find-debug-server.cpp#L534-L539

If you're setting `QUICK_LINT_JS_FEATURE_DEBUG_SERVER`, then I can see other things failing, but not `find_debug_servers`.

**Review by strager — COMMENTED** [review]
> Ideally this file shouldn't get compiled at all when the debug server
feature is disabled.

No, it should always be compiled. Non-debug-server builds should be able to find debug servers.

File: src/quick-lint-js/debug/find-debug-server.cpp [hunk 1]
```diff
@@ -434,6 +470,83 @@ std::vector<found_debug_server> find_debug_servers() {
...
+  if (rc < 0) {
+    QLJS_DEBUG_LOG("%s: ignoring failure to query own jail id: %s\n", func,
+                   ::strerror(errno));
+    return {};
+  }
+
+  // Open KVM access device
+  kd = ::kvm_openfiles(nullptr /* execfile */, "/dev/null" /* corefile */,
+                       nullptr /* swapfile (unused) */, O_RDONLY /* flags */,
+                       errbuf);
```
Comment (strager) [inline src/quick-lint-js/debug/find-debug-server.cpp:511]: Nit:
```suggestion
  kd = ::kvm_openfiles(/*execfile=*/nullptr, /*corefile=*/"/dev/null",
                       /*swapfile=*/nullptr, /*flags=*/O_RDONLY,
                       errbuf);
```

**Review by strager — APPROVED** [review]
Patch looks good to me. I plan on squashing the two commits.

Do the relevant tests in test/test-debug-server.cpp work? In particular: `./build-native/test/quick-lint-js-test --gtest_filter=test_debug_server.find_debug_servers_finds_running_instance_SLOW`

**Comment by herrhotzenplotz** [comment]
> > Ideally this file shouldn't get compiled at all when the debug server
> > feature is disabled.
> 
> No, it should always be compiled. Non-debug-server builds should be able to find debug servers.

Ah ok, that makes sense to me.

**Comment by herrhotzenplotz** [comment]
> Patch looks good to me. I plan on squashing the two commits.
> 
> Do the relevant tests in test/test-debug-server.cpp work? In particular: `./build-native/test/quick-lint-js-test --gtest_filter=test_debug_server.find_debug_servers_finds_running_instance_SLOW`

Yes, I tested on arm64 and amd64. Both pass perfectly fine.

File: src/quick-lint-js/debug/find-debug-server.cpp [hunk 2]
```diff
@@ -191,6 +218,15 @@ void register_current_thread_as_debug_server_thread(std::uint16_t port_number) {
     return;
   }
 #endif
+#if defined(__FreeBSD__)
+  int rc = ::pthread_setname_np(pthread_self(), name.data());
+  if (rc != 0) {
+    QLJS_DEBUG_LOG(
+        "%s: ignoring failure to set thread name for debug server thread: %s\n",
+        __func__, std::strerror(errno));
```
[the remark this answers is not in the document]
Comment (herrhotzenplotz) [reply src/quick-lint-js/debug/find-debug-server.cpp:226]: Good point. The manual page doesn't mention it but the source code clearly returns the error through the return value and not errno.

**Review by strager — APPROVED** [review]

**Pull request merged**